### PR TITLE
feat: Migrate to Realm

### LINKED ISSUE #1466: SQLite replacement ORM library
**Issue**

SQLite database is slow, hard to manage, hard to test and terse with lot joins and requires incredible amount of boilerplate code to create simple use cases

**What should be done?**

Replace SQlite with any of the modern ORM wrapper or for better efficacy, a different database solution overall, which provides:
- reactive notifications about data changes as an event application should be as reactive as possible
- easy to test as we would want to mock the database easily to create unit tests. Our goal should be to abstract the data layer behind repository pattern so that it is easily testable
- easy to maintain as the open event API keeps changing, so we should be able to just change our object models and not worry about the app breaking. So either automatic migrations or a solution where it is easy to migrate data should be preferred
- requires less maintenance as only maintaining the production is hard enough and we won't want to maintain the boilerplate as well
- requires less time to implement

**Strategy**

Discuss about your preferences and their pros and cons and when it is decided, one person chooses to create a branch on his/her fork and other developers collaborate with him/her and complete the implementation as soon as possible.  

**Milestone**
[Milestone 1](https://github.com/fossasia/open-event-android/milestone/24)

### PR DESCRIPTION
**Pull request by iamareebjamal**
Fixes #1466

### CODE REVIEW

**Comment by iamareebjamal** [comment]
Current Bugs: 
- Schedule Filtering not working correctly
- Sort By Tracks not working ( Realm does not support sorting by Primary Key. Either sort in memory or maintain clone field )
- Bookmark change notifications are pushed but UI does not refresh correctly sometimes. ( Fix UI refresh rather than statically setting bookmark icons )
- Weird menu icons in ScheduleFragment ( maybe a bug in master which wasn't apparent because of slow query )
- Wrong mapping of Track Color ( Preferably update models to include Track information )

**Comment by iamareebjamal** [comment]
Everything is complete.Please test @the-dagger @mananwason 

@arp95 @free4murad @harshitagupta30 @manikaw @Shailes351

Will need your help in sorting out issues if there are some. My exams are starting soon, so I have done most of the work in these 2 days

Any help will be much appreciated
Thank you

**Comment by free4murad** [comment]
Tried running the app. Looks good. (Minor issue related to search, but I think we can merge this)
[Here](https://drive.google.com/file/d/0ByRo_SG_Si40eHpIeXh5Zzg2clU/view?usp=sharing) is the APK.
Please install and share your reviews

**Comment by iamareebjamal** [comment]
@free4murad What was the issue?

**Comment by iamareebjamal** [comment]
People, need reviews! The more commits added to dev branch, more difficult it will be to rebase

**Comment by the-dagger** [comment]
@iamareebjamal while it's quite hard to review a large PR, what we can do is generate the apk and test it out.
If that works without issues then we can go ahead and merge it.
Sounds good?

**Comment by iamareebjamal** [comment]
@the-dagger Yeah, definitely

Thank you :smile:

**Comment by mahikaw** [comment]
Agreed with @the-dagger.  I too will run it once and check thoroughly. 
Also, @iamareebjamal I'd suggest you to squash them once you're done with it, so that even after merging if we face any kind of problem, we could simply revert the commit and we'll be good to go. 😄

**Comment by iamareebjamal** [comment]
Yes, I will squash when changes are approved. They are this way because if I want to temporarily remove something or revert an operation if not approved easily

File: android/app/src/main/java/org/fossasia/openevent/OpenEventApp.java [hunk 1]
```diff
@@ -95,6 +97,12 @@ public void onCreate() {
         sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getAppContext());
         sDefSystemLanguage = Locale.getDefault().getDisplayLanguage();
 
+        RealmConfiguration config = new RealmConfiguration.Builder()
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/OpenEventApp.java]: Thanks for the catch

**Comment by iamareebjamal** [comment]
@Shailesh351 

- Fixed Bookmark crash
- Fixed Realm init crash
- Fixed relation error
- Sorting was working fine

Filtering is delegated as this is essentially adding realm and current filtering process is very inefficient and causes memory leaks and a laggy UI

If you want, you can add a PR to this branch fixing the filtering

**Comment by mananwason** [comment]
@iamareebjamal Let's add filtering to this PR as well. We want the app to be in the previous state before we can merge all these commits into development.

**Comment by iamareebjamal** [comment]
I'm currently busy in exams but I'll try my best to include it ASAP

**Comment by Shailesh351** [comment]
**Issues :** 

1. Bookmark fragment : After bookmarking sessions when removing bookmark from bookmarks fragment sometimes getting following error 
`D/BookmarksFragment: Bookmarks changed
05-13 10:27:11.217 11386-11386/org.fossasia.openevent I/Choreographer: Skipped 35 frames!  The application may be doing too much work on its main thread.
05-13 10:27:13.019 11386-11386/org.fossasia.openevent D/BookmarksFragment: Bookmarks changed
05-13 10:27:13.019 11386-11386/org.fossasia.openevent I/Choreographer: Skipped 54 frames!  The application may be doing too much work on its main thread.
05-13 10:27:32.407 11386-11386/org.fossasia.openevent D/BookmarksFragment: Bookmarks changed
05-13 10:27:34.145 11386-11386/org.fossasia.openevent D/BookmarksFragment: Bookmarks changed
05-13 10:27:34.146 11386-11386/org.fossasia.openevent I/Choreographer: Skipped 50 frames!  The application may be doing too much work on its main thread.
05-13 10:27:39.666 11386-11386/org.fossasia.openevent D/BookmarksFragment: Bookmarks changed
05-13 10:27:47.972 11386-11386/org.fossasia.openevent D/SessionDetailActivity: Updated
05-13 10:27:47.973 11386-11386/org.fossasia.openevent D/Session Detail: Bookmark Removed
                                                               beginning of crash
05-13 10:27:47.974 11386-11386/org.fossasia.openevent A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x0 in tid 11386 (sasia.openevent)
                                                              [ 05-13 10:27:47.975   387:  387 W/]
                                                              debuggerd: handling request: pid=11386 uid=10420 gid=10420 tid=11386`

2. Sometimes after removing all bookmarked sessions dialogue appears. After clicking ''Ok'' same dialogue appears and on clicking that "Ok" app is crashing with error message
`java.lang.NullPointerException: Attempt to invoke virtual method 'android.support.v4.app.FragmentTransaction android.support.v4.app.FragmentManager.beginTransaction()' on a null object reference
                                                                            at org.fossasia.openevent.fragments.BookmarksFragment$1.onClick(BookmarksFragment.java:80)` 

3. Sessions in tracks are not sorted according to time

**Comment by iamareebjamal** [comment]
Should be fixed now @Shailesh351

**Comment by iamareebjamal** [comment]
Native code implementation of Realm was causing Segmentation Fault. Should be fixed now

**Comment by Shailesh351** [comment]
Everything works fine except filtering in day schedule fragment :smile: :+1:

**Comment by iamareebjamal** [comment]
@mananwason Please consider this library to be added, it'll solve half of our problems https://github.com/davideas/FlexibleAdapter

**Comment by iamareebjamal** [comment]
Phew, done! Thanks to @m-murad, his commit and some modifications fixed the filter.

Please test @Shailesh351 @m-murad @manikaw @arp95 so we can merge this once and for all

**Comment by Shailesh351** [comment]
@iamareebjamal When page is switched in view pager and sort option is changed headers are not arranged properly.

Sent PR for fixing this on your fork

**Comment by arp95** [comment]
@iamareebjamal
nice work!
outside right now. will check by night.

**Comment by iamareebjamal** [comment]
@Shailesh351 You didn't call notifySort after the options are changed, so it wasn't showing correct order. Can you test now please?

**Comment by Shailesh351** [comment]
@iamareebjamal Actually i changed it because when filter is already applied and then sort is changed then sessions are not getting filtered

**Comment by m-murad** [comment]
@iamareebjamal 
Is it working fine now or do we need to fix something?

Please tell so that I can send a PR.

**Comment by iamareebjamal** [comment]
@m-murad Can you please test if everything is working now?

**Comment by iamareebjamal** [comment]
@Shailesh351 Sorting and Filtering working fine now?

**Comment by m-murad** [comment]
@iamareebjamal 
When we filter sessions and change the page on viewpageradapter the sessions are not filtered. 

Solving this. Will ping you once I am done.

**Comment by arp95** [comment]
@iamareebjamal 
working fine. just check the filtering part thats it i think.

**Comment by Shailesh351** [comment]
@iamareebjamal Resolve conflict . I think this branch is now equal to development. :+1:

**Comment by iamareebjamal** [comment]
Resolving conflicts @Shailesh351 but there is still a bug in filtering. @m-murad is working on it

**Comment by iamareebjamal** [comment]
@m-murad has fixed the filter :tada: 

@Shailesh351 @arp95 @mahikaw Please test the app now

**Comment by rohanagarwal94** [comment]
@iamareebjamal firstly great work. There are conflicts. Please resolve them before it could be reviewed. Cheers!

**Comment by iamareebjamal** [comment]
@rohanagarwal94 Surely. A little busy in exams, but tomorrow is my last one. Will do it soon

**Comment by iamareebjamal** [comment]
@rohanagarwal94 @arp95 @mahikaw @heysadboy @Shailesh351 Please test the apk before it breaks again :rofl: 
 
https://www.dropbox.com/s/dhipvfrs1beroy6/app-fdroid-debug.apk?dl=1

**Comment by heysadboy** [comment]
App crashes when I open it after installing see, this is the [log](https://drive.google.com/file/d/0B6BdphmVRuujbER3YlFrRkdUQzA/view?usp=sharing),

screenshot from 2017-06-01 04-41-46

**Comment by iamareebjamal** [comment]
Did you delete the old app first and reboot? New app is almost a complete rewrite so if you were using Instant Run, this is normal. Please delete, reboot and install

I'll look into it nevertheless

**Comment by the-dagger** [comment]
@iamareebjamal is this good to merge?

**Comment by iamareebjamal** [comment]
@the-dagger I have tested everything and seems to work fine. I have asked others to review as well. But it is good to merge AFAIK

**Comment by the-dagger** [comment]
Alright, let's wait for one more feedback before merging this.
Also could you resolve the conflict in MapFragment?

**Comment by rohanagarwal94** [comment]
@iamareebjamal the app is crashing when I tried opening the attached apk.

screenshot_2017-06-01-22-46-35-208_com miui bugreport

screenshot_2017-06-01-22-46-40-665_com miui bugreport

**Comment by iamareebjamal** [comment]
@rohanagarwal94 and Everyone. Please ensure that you don't update the app, but uninstall it and then reboot and then reinstall. This is a common issue if you use Instant Run

This is version of app has a very different bytecode than the one installed via instant run, it is going to throw issues

**Comment by rohanagarwal94** [comment]
I downloaded the apk through the usb, uninstalled the existing app, installed the downloaded apk. Correct me where I could be wrong.

**Comment by iamareebjamal** [comment]
When you don't reboot, the dalvik cache can interfere in debug builds with significant code changes

Nevertheless I am uploading new APKs, these should work without reboot. Made clean builds without instant run

**Comment by iamareebjamal** [comment]
Google Play Variant : https://www.dropbox.com/s/1ji7nagw9zd40dt/app-googleplay-debug.apk?dl=1

**Comment by Shailesh351** [comment]
@iamareebjamal 
1) When first time running app and user selects "No" in dialogue app closes suddenly  only once and data is not loaded from assets
`Caused by: java.lang.ClassCastException: org.fossasia.openevent.data.Track cannot be cast to org.fossasia.openevent.data.Session
                                                                            at org.fossasia.openevent.activities.MainActivity.lambda$handleJsonEvent$6(MainActivity.java:781)
                                                                            at org.fossasia.openevent.activities.MainActivity$$Lambda$6.run(Unknown Source)
                                                                            at io.reactivex.internal.operators.completable.CompletableFromAction.subscribeActual(CompletableFromAction.java:34)
                                                                            at io.reactivex.Completable.subscribe(Completable.java:1634)
                                                                            at io.reactivex.internal.operators.completable.CompletableObserveOn.subscribeActual(CompletableObserveOn.java:34)
                                                                            at io.reactivex.Completable.subscribe(Completable.java:1634)
                                                                            at io.reactivex.Completable.subscribe(Completable.java:1622)
                                                                            at org.fossasia.openevent.activities.MainActivity.handleJsonEvent(MainActivity.java:815)
                                                                            at java.lang.reflect.Method.invoke(Native Method)
                                                                            at com.squareup.otto.EventHandler.handleEvent(EventHandler.java:89)
                                                                            at com.squareup.otto.Bus.dispatch(Bus.java:385)
                                                                            at com.squareup.otto.Bus.dispatchQueuedEvents(Bus.java:368)
                                                                            at com.squareup.otto.Bus.post(Bus.java:337)
                                                                            at org.fossasia.openevent.OpenEventApp.lambda$postEventOnUIThread$0(OpenEventApp.java:71)`

**Comment by heysadboy** [comment]
@iamareebjamal This one is working. Great work!

**Comment by rohanagarwal94** [comment]
@iamareebjamal the markers are not visible in the google play flavour.

**Comment by iamareebjamal** [comment]
@rohanagarwal94 As you worked on the PR, can you please take a look

**Comment by iamareebjamal** [comment]
@Shailesh351 Fixed. Please test now

**Comment by rohanagarwal94** [comment]
debugging it out locally. Although didnt find anything just by glancing on the new codebase.

**Comment by iamareebjamal** [comment]
Yeah. I just resolved the conflicts from your PR and pushed. Should've worked correctly. 

Thanks for the help

**Comment by the-dagger** [comment]
Ok, merging this then.

**Comment by iamareebjamal** [comment]
@the-dagger The map issue wasn't resolved, I was just telling @rohanagarwal94 that I just resolved conflicts from his build and it shouldn't have broken things. I'll create a PR for that issue now tomorrow

Sorry for the confusion

**Comment by rohanagarwal94** [comment]
Never mind let me create a separate PR. I have resolved it.

**Pull request merged**